feat(Tag): Implement manage tags dialog

geoffreykwan opened the pull request:
## Changes
- Implemented manage tags dialog
- Tag text can now be edited

## Test
- Test with https://github.com/WISE-Community/WISE-API/pull/261
- In the Class Schedule view, select some units by clicking on their checkbox
- Click the "Apply Tags" button
- Click the "Manage Tags" menu item
- Change the text for a tag. The tag text should immediately update in the "Apply tags" menu but not in the unit cards. I'll fix immediately updating in the unit cards in another PR.

geoffreykwan committed:
feat(Tag): Implement manage tags dialog

geoffreykwan committed:
chore(Translation): Generate translations

hirokiterashima commented on src/app/teacher/apply-tags-button/apply-tags-button.component.ts:
```diff
@@ -21,6 +23,10 @@ export class ApplyTagsButtonComponent implements OnInit {
+    this.tagService.tagChanged$.subscribe((tagThatChanged: Tag) => {
+      const tag = this.tags.find((t: Tag) => t.id === tagThatChanged.id);
+      tag.text = tagThatChanged.text;
+    });
```
Does this need to be unsubscribed in ```ngOnDestroy()```? Also, would ```updatedTag``` be a better name here?

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts:
```diff
@@ -0,0 +1,53 @@
+  constructor(private tagService: TagService) {}
+
+  ngOnInit(): void {
+    this.tagService.retrieveUserTags().subscribe((tags: Tag[]) => {
```
unsubscribe in onDestroy()?

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts:
```diff
@@ -0,0 +1,53 @@
+  ]
+})
+export class ManageTagsDialogComponent implements OnInit {
+  inputChanged: Subject<string> = new Subject<string>();
```
protected?

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts:
```diff
@@ -0,0 +1,53 @@
+})
+export class ManageTagsDialogComponent implements OnInit {
+  inputChanged: Subject<string> = new Subject<string>();
+  inputChangedSubscription: Subscription;
```
private?

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.ts:
```diff
@@ -0,0 +1,53 @@
+    this.inputChangedSubscription = this.inputChanged
+      .pipe(debounceTime(1000), distinctUntilChanged())
+      .subscribe(({ tag }: any) => {
+        this.updateTag(tag);
```
Call service function directly and remove need to define a new function?
```suggestion
        this.tagService.updateTag(tag);
```

hirokiterashima commented on src/app/teacher/manage-tags-dialog/manage-tags-dialog.component.html:
```diff
@@ -0,0 +1,18 @@
+          matInput
+          type="text"
+          [(ngModel)]="tag.text"
+          (ngModelChange)="inputChanged.next({ tag: tag, text: $event })"
```
```ManageTagsDialogComponent.inputChanged``` has type ```Subject<string>```, but here we're passing in an object. What should it be?

hirokiterashima commented on src/assets/wise5/services/tagService.ts:
```diff
@@ -100,4 +102,10 @@ export class TagService {
     return this.http.delete(`/api/projects/tag/${tag.text}`, { params: params }).subscribe();
   }
+
+  updateTag(tag: Tag): Subscription {
+    return this.http.put<Tag>(`/api/user/tag/${tag.id}`, tag).subscribe((tag) => {
+      this.tagChangedSource.next(tag);
+    });
+  }
```
Do we need to return a Subscription? It looks like we don't do anything with it?
```suggestion
  updateTag(tag: Tag): void {
    this.http.put<Tag>(`/api/user/tag/${tag.id}`, tag).subscribe((tag) => {
      this.tagChangedSource.next(tag);
    });
  }
```

The other option is to return an Observable<Tag> and subscribe in the component.
```
  updateTag(tag: Tag): Observable<Tag> {
    return this.http.put<Tag>(`/api/user/tag/${tag.id}`, tag).pipe(
      tap((tag) => {
        this.tagChangedSource.next(tag);
      })
    );
  }
```

hirokiterashima reviewed (changes requested):
Functionality works as described. I have a few comments:

1. We discussed using a snack bar to indicate that the change has been saved.
2. I added some code change suggestions inline.

I'm also wondering if we should split up the TagService to StudentTagService and UserTagService. It looks like there are two different kinds of tags right now, and handling them both in TagService is confusing. This refactoring can be done in a separate PR. Let's discuss how best to proceed.

geoffreykwan committed:
refactor(Tag): Clean up code

- Unsubscribe from subscriptions
- Rename some variables
- Add protected and private modifiers

geoffreykwan committed:
feat(Tag): Show snack bar message when tag is updated

geoffreykwan commented:
Ok, I've made the changes.

hirokiterashima reviewed (approved):
LGTM

In ManageTagsDialogComponent, 
```protected inputChanged: Subject<string> = new Subject<any>();```
should be 
```protected inputChanged: Subject<any> = new Subject<any>();```
?

geoffreykwan committed:
chore(Tag): Fix variable using wrong type
